src/Common/src/System/Runtime/Serialization/SerializationGuard.cs

```diff
@@ -0,0 +1,54 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Reflection;
+
+namespace System.Runtime.Serialization
+{
+    /// <summary>
+    /// Provides access to portions of the Serialization Guard APIs since they're not publicly exposed via contracts.
+    /// </summary>
+    internal static partial class SerializationGuard
+    {
+        private delegate void ThrowIfDeserializationInProgressWithSwitchDel(string switchName, ref int cachedValue);
+        private static readonly ThrowIfDeserializationInProgressWithSwitchDel s_throwIfDeserializationInProgressWithSwitch = CreateThrowIfDeserializationInProgressWithSwitchDelegate();
+
+        /// <summary>
+        /// Equivalent to typeof(DeserializationBlockedException).
+        /// </summary>
+        public static readonly Type DeserializationBlockedExceptionType = GetDeserializationBlockedExceptionType();
+
+        /// <summary>
+        /// Builds a wrapper delegate for SerializationInfo.ThrowIfDeserializationInProgress(string, ref int),
+        /// since it is not exposed via contracts.
+        /// </summary>
+        private static ThrowIfDeserializationInProgressWithSwitchDel CreateThrowIfDeserializationInProgressWithSwitchDelegate()
+        {
+            ThrowIfDeserializationInProgressWithSwitchDel throwIfDeserializationInProgressDelegate = null;
+            MethodInfo throwMethod = typeof(SerializationInfo).GetMethod("ThrowIfDeserializationInProgress",
+                BindingFlags.Static | BindingFlags.NonPublic | BindingFlags.Public, null, new Type[] { typeof(string), typeof(int).MakeByRefType() }, new ParameterModifier[0]);
+
+            if (throwMethod != null)
+            {
+                throwIfDeserializationInProgressDelegate = (ThrowIfDeserializationInProgressWithSwitchDel)throwMethod.CreateDelegate(typeof(ThrowIfDeserializationInProgressWithSwitchDel));
+            }
+
+            return throwIfDeserializationInProgressDelegate;
+        }
+
+        private static Type GetDeserializationBlockedExceptionType()
+        {
+            return typeof(SerializationInfo).Assembly.GetType("System.Runtime.Serialization.DeserializationBlockedException", throwOnError: false);
```

Comment:
Can we just use a regular `SerializationException` for this, and get rid of the DeserializationBlockedException completely?

Reply:
That would change the control flow of some callers. For example, `ObjectManager` special-cases the `DeserializationBlockedException` type and rethrows it rather than allowing a typical `TargetInvocationException` to escape.

https://github.com/dotnet/corefx/blob/9a0633c60d62d227386a87e66ee2b7ebada771e7/src/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/ObjectManager.cs#L780-L785

In order to preserve this runtime behavior, would it make sense for us to leave `DeserializationBlockedException` in (but not exposed via contracts), but have it subclass `SerializationException` instead of the base `Exception` type?

Reply:
Why is this runtime behavior important? There are number of other places that will wrap the exception.

Reply:
(Correction to my above comment: it seems `DeserializationBlockedException` already subclasses `SerializationException` in the implementation. The ref assembly incorrectly shows it as subclassing `Exception` directly.)

I wouldn't call the existing behavior "important" _per se_, but I was trying to avoid behavioral changes in this PR. If consensus is that reusing `SerializationException` directly is the preferred action, and if the behavioral change isn't concerning, we can absolutely piggy-back that on this PR.